File: frontend/lib/core/models/reminder.dart
enum ReminderType { medicine, hydration, activity, appointment }

enum AckStatus { acknowledged, missed }

extension ReminderTypeLabel on ReminderType {
  String get label => switch (this) {
        ReminderType.medicine => 'Medicine',
        ReminderType.hydration => 'Hydration',
        ReminderType.activity => 'Daily activity',
        ReminderType.appointment => 'Appointment',
      };
}

class Reminder {
  final String id;
  final String patientId;

  /// A caregiver_id, or an ASHA acting as fallback caregiver-of-record.
  final String createdBy;
  final ReminderType type;
  final String schedule;

  /// Null falls back to TTS; recorded family-voice clips are the headline feature.
  final String? voiceClipUrl;
  final String textLabel;

  const Reminder({
    required this.id,
    required this.patientId,
    required this.createdBy,
    required this.type,
    required this.schedule,
    required this.textLabel,
    this.voiceClipUrl,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) => Reminder(
        id: json['id'] as String,
        patientId: json['patient_id'] as String,
        createdBy: json['created_by'] as String,
        type: ReminderType.values.byName(json['type'] as String),
        schedule: json['schedule'] as String,
        voiceClipUrl: json['voice_clip_url'] as String?,
        textLabel: json['text_label'] as String,
      );

  Map<String, dynamic> toJson() => {
        'patient_id': patientId,
        'created_by': createdBy,
        'type': type.name,
        'schedule': schedule,
        'voice_clip_url': voiceClipUrl,
        'text_label': textLabel,
      };
}

class ReminderAck {
  final String reminderId;
  final String patientId;
  final DateTime timestamp;
  final AckStatus status;

  const ReminderAck({
    required this.reminderId,
    required this.patientId,
    required this.timestamp,
    required this.status,
  });

  Map<String, dynamic> toJson() => {
        'reminder_id': reminderId,
        'patient_id': patientId,
        'timestamp': timestamp.toIso8601String(),
        'status': status.name,
      };

  factory ReminderAck.fromJson(Map<String, dynamic> json) => ReminderAck(
        reminderId: json['reminder_id'] as String,
        patientId: json['patient_id'] as String,
        timestamp: DateTime.parse(json['timestamp'] as String),
        status: AckStatus.values.byName(json['status'] as String),
      );
}

File: backend/app/models/reminder.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class ReminderType(str, enum.Enum):
    medicine = "medicine"
    hydration = "hydration"
    activity = "activity"
    appointment = "appointment"


class AckStatus(str, enum.Enum):
    acknowledged = "acknowledged"
    missed = "missed"


class Reminder(Base):
    """created_by may be a caregiver_id or an asha_id acting as fallback caregiver-of-record."""

    __tablename__ = "reminders"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    patient_id: Mapped[str] = mapped_column(ForeignKey("patients.id"), nullable=False)
    created_by: Mapped[str] = mapped_column(ForeignKey("users.id"), nullable=False)
    type: Mapped[ReminderType] = mapped_column(Enum(ReminderType), nullable=False)
    schedule: Mapped[str] = mapped_column(String, nullable=False)  # cron-like or ISO recurrence string
    voice_clip_url: Mapped[str | None] = mapped_column(String, nullable=True)  # falls back to TTS if null
    text_label: Mapped[str] = mapped_column(String, nullable=False)


class ReminderAck(Base):
    __tablename__ = "reminder_acks"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    reminder_id: Mapped[str] = mapped_column(ForeignKey("reminders.id"), nullable=False)
    patient_id: Mapped[str] = mapped_column(ForeignKey("patients.id"), nullable=False)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[AckStatus] = mapped_column(Enum(AckStatus), nullable=False)
